File: src/test/java/network/crypta/node/VersionTest.java
package network.crypta.node;

import org.junit.jupiter.api.Test;

import static network.crypta.node.Version.*;
import static org.junit.jupiter.api.Assertions.*;

/**
 * Unit tests for {@link Version}, focusing on compareBuildNumbers(), isBuildAtLeast(), and
 * parseNodeNameFromVersionStr().
 */
public class VersionTest {

  // Constants for testing with the same values exposed by Version.
  private static final String CRYPTAD_NODE_NAME = "Cryptad";
  private static final String FRED_NODE_NAME = "Fred";
  private static final int MIN_CRYPTAD_BUILD = 1;
  private static final int MIN_FRED_BUILD = 1475;

  // Test version strings
  private static final String VALID_CRYPTAD_VERSION = "Cryptad,1504,1.0,1504";
  private static final String VALID_FRED_VERSION = "Fred,0.7,1.0,1503";
  private static final String MINIMAL_CRYPTAD_VERSION = "Cryptad,1,1.0";
  private static final String MINIMAL_FRED_VERSION = "Fred,0.7,1.0,1";
  private static final String MALFORMED_VERSION_SHORT = "Cryptad,1504";
  private static final String MALFORMED_VERSION_EMPTY = "";
  private static final String MALFORMED_VERSION_WRONG_SEPARATOR = "Cryptad;1504;1.0";

  @Test
  public void testCompareBuildNumbers_CryptadVsFred() {
    // Cryptad should always be considered newer than Fred regardless of build numbers
    assertTrue(
        compareBuildNumbers(CRYPTAD_NODE_NAME, 1, FRED_NODE_NAME, 9999) > 0,
        "Cryptad should be newer than Fred with lower build number");
    assertTrue(
        compareBuildNumbers(CRYPTAD_NODE_NAME, 1500, FRED_NODE_NAME, 1500) > 0,
        "Cryptad should be newer than Fred with same build number");
    assertTrue(
        compareBuildNumbers(CRYPTAD_NODE_NAME, 9999, FRED_NODE_NAME, 1) > 0,
        "Cryptad should be newer than Fred with higher build number");

    // Fred should be older than Cryptad
    assertTrue(
        compareBuildNumbers(FRED_NODE_NAME, 1, CRYPTAD_NODE_NAME, 9999) < 0,
        "Fred should be older than Cryptad with lower build number");
    assertTrue(
        compareBuildNumbers(FRED_NODE_NAME, 1500, CRYPTAD_NODE_NAME, 1500) < 0,
        "Fred should be older than Cryptad with same build number");
    assertTrue(
        compareBuildNumbers(FRED_NODE_NAME, 9999, CRYPTAD_NODE_NAME, 1) < 0,
        "Fred should be older than Cryptad with higher build number");
  }

  @Test
  public void testCompareBuildNumbers_SameNodeType() {
    // Same node type - should compare build numbers numerically

    // Cryptad vs Cryptad
    assertTrue(
        compareBuildNumbers(CRYPTAD_NODE_NAME, 1505, CRYPTAD_NODE_NAME, 1504) > 0,
        "Higher Cryptad build should be newer");
    assertTrue(
        compareBuildNumbers(CRYPTAD_NODE_NAME, 1504, CRYPTAD_NODE_NAME, 1505) < 0,
        "Lower Cryptad build should be older");
    assertEquals(
        0,
        compareBuildNumbers(CRYPTAD_NODE_NAME, 1504, CRYPTAD_NODE_NAME, 1504),
        "Same Cryptad build should be equal");

    // Fred vs Fred
    assertTrue(
        compareBuildNumbers(FRED_NODE_NAME, 1503, FRED_NODE_NAME, 1502) > 0,
        "Higher Fred build should be newer");
    assertTrue(
        compareBuildNumbers(FRED_NODE_NAME, 1502, FRED_NODE_NAME, 1503) < 0,
        "Lower Fred build should be older");
    assertEquals(
        0,
        compareBuildNumbers(FRED_NODE_NAME, 1503, FRED_NODE_NAME, 1503),
        "Same Fred build should be equal");
  }

  @Test
  public void testCompareBuildNumbers_NullNodeNames() {
    // When node names are null, should fall back to build number comparison
    assertTrue(
        compareBuildNumbers(null, 1505, null, 1504) > 0,
        "Higher build number should win with null node names");
    assertTrue(
        compareBuildNumbers(null, 1504, null, 1505) < 0,
        "Lower build number should lose with null node names");
    assertEquals(
        0,
        compareBuildNumbers(null, 1504, null, 1504),
        "Same build number should be equal with null node names");

    // One null, one not null - should fall back to build number comparison
    assertTrue(
        compareBuildNumbers(null, 1505, FRED_NODE_NAME, 1504) > 0,
        "Higher build number should win with one null node name");
    assertTrue(
        compareBuildNumbers(CRYPTAD_NODE_NAME, 1504, null, 1505) < 0,
        "Lower build number should lose with one null node name");
  }

  @Test
  public void testCompareBuildNumbers_EdgeCases() {
    // Test with zero and negative build numbers
    assertTrue(
        compareBuildNumbers(CRYPTAD_NODE_NAME, 1, CRYPTAD_NODE_NAME, 0) > 0,
        "Positive should be greater than zero");
    assertTrue(
        compareBuildNumbers(FRED_NODE_NAME, 0, FRED_NODE_NAME, -1) > 0,
        "Zero should be greater than negative");

    // Test with very large build numbers
    assertTrue(
        compareBuildNumbers(
                CRYPTAD_NODE_NAME, Integer.MAX_VALUE, CRYPTAD_NODE_NAME, Integer.MAX_VALUE - 1)
            > 0,
        "Large build number comparison should work");
  }

  @Test
  public void testCompareBuildNumbers_UnknownNodeTypes() {
    // Unknown node types should fall back to build number comparison
    assertTrue(
        compareBuildNumbers("Unknown", 1505, "AnotherUnknown", 1504) > 0,
        "Unknown node type should compare by build number");
    assertTrue(
        compareBuildNumbers("Unknown", 1505, "SomeOther", 1504) > 0,
        "Unknown vs known should compare by build number when not Cryptad/Fred");
  }

  @Test
  public void testIsBuildAtLeast_CryptadNode() {
    // Cryptad nodes should always meet Fred minimum requirements
    assertTrue(
        isBuildAtLeast(CRYPTAD_NODE_NAME, 1, MIN_FRED_BUILD),
        "Cryptad should always meet Fred minimum build requirement");
    assertTrue(
        isBuildAtLeast(CRYPTAD_NODE_NAME, 1, 9999),
        "Cryptad with low build should still meet Fred minimum");
    assertTrue(
        isBuildAtLeast(CRYPTAD_NODE_NAME, 0, MIN_FRED_BUILD),
        "Cryptad with zero build should still meet Fred minimum");
    assertTrue(
        isBuildAtLeast(CRYPTAD_NODE_NAME, -1, MIN_FRED_BUILD),
        "Cryptad with negative build should still meet Fred minimum");
  }

  @Test
  public void testIsBuildAtLeast_FredNode() {
    // Fred nodes should be checked against the minimum build number
    assertTrue(
        isBuildAtLeast(FRED_NODE_NAME, MIN_FRED_BUILD, MIN_FRED_BUILD),
        "Fred node with sufficient build should meet minimum");
    assertTrue(
        isBuildAtLeast(FRED_NODE_NAME, MIN_FRED_BUILD + 1, MIN_FRED_BUILD),
        "Fred node with higher build should meet minimum");
    assertFalse(
        isBuildAtLeast(FRED_NODE_NAME, MIN_FRED_BUILD - 1, MIN_FRED_BUILD),
        "Fred node with insufficient build should not meet minimum");
    assertFalse(
        isBuildAtLeast(FRED_NODE_NAME, 0, MIN_FRED_BUILD),
        "Fred node with zero build should not meet minimum");
  }

  @Test
  public void testIsBuildAtLeast_NullAndUnknownNodeNames() {
    // Null node names should be treated like Fred nodes
    assertTrue(
        isBuildAtLeast(null, MIN_FRED_BUILD, MIN_FRED_BUILD),
        "Null node name with sufficient build should meet minimum");
    assertFalse(
        isBuildAtLeast(null, MIN_FRED_BUILD - 1, MIN_FRED_BUILD),
        "Null node name with insufficient build should not meet minimum");

    // Unknown node names should be treated like Fred nodes
    assertTrue(
        isBuildAtLeast("UnknownNode", MIN_FRED_BUILD, MIN_FRED_BUILD),
        "Unknown node name with sufficient build should meet minimum");
    assertFalse(
        isBuildAtLeast("UnknownNode", MIN_FRED_BUILD - 1, MIN_FRED_BUILD),
        "Unknown node name with insufficient build should not meet minimum");
  }

  @Test
  public void testIsBuildAtLeast_EdgeCases() {
    // Test boundary conditions
    assertTrue(isBuildAtLeast(FRED_NODE_NAME, 1000, 1000), "Exact minimum should meet requirement");
    assertFalse(
        isBuildAtLeast(FRED_NODE_NAME, 999, 1000), "One below minimum should not meet requirement");
    assertTrue(
        isBuildAtLeast(FRED_NODE_NAME, 1001, 1000), "One above minimum should meet requirement");

    // Test with zero minimum
    assertTrue(isBuildAtLeast(FRED_NODE_NAME, 0, 0), "Any build should meet zero minimum");
    assertTrue(isBuildAtLeast(FRED_NODE_NAME, 1, 0), "Positive build should meet zero minimum");
    assertFalse(
        isBuildAtLeast(FRED_NODE_NAME, -1, 0), "Negative build should not meet zero minimum");
  }

  @Test
  public void testParseNodeNameFromVersionStr_ValidVersions() {
    // Test valid Cryptad version strings
    assertEquals(
        CRYPTAD_NODE_NAME,
        parseNodeNameFromVersionStr(VALID_CRYPTAD_VERSION),
        "Should extract Cryptad node name");
    assertEquals(
        CRYPTAD_NODE_NAME,
        parseNodeNameFromVersionStr(MINIMAL_CRYPTAD_VERSION),
        "Should extract Cryptad from minimal version");

    // Test valid Fred version strings
    assertEquals(
        FRED_NODE_NAME,
        parseNodeNameFromVersionStr(VALID_FRED_VERSION),
        "Should extract Fred node name");
    assertEquals(
        FRED_NODE_NAME,
        parseNodeNameFromVersionStr(MINIMAL_FRED_VERSION),
        "Should extract Fred from minimal version");

    // Test single component version (edge case)
    assertEquals(
        CRYPTAD_NODE_NAME,
        parseNodeNameFromVersionStr("Cryptad"),
        "Should extract node name from single component");
    assertEquals(
        FRED_NODE_NAME,
        parseNodeNameFromVersionStr("Fred"),
        "Should extract node name from single component");
  }

  @Test
  public void testParseNodeNameFromVersionStr_NullAndEmptyInputs() {
    // Test null input
    assertNull(parseNodeNameFromVersionStr(null), "Null version string should return null");

    // Test empty input - Fields.commaList("") returns empty array, so we get null
    assertNull(
        parseNodeNameFromVersionStr(MALFORMED_VERSION_EMPTY),
        "Empty version string should return null");

    // Test whitespace-only input - Fields.commaList("   ") returns [""] (one empty string)
    assertEquals(
        "",
        parseNodeNameFromVersionStr("   "),
        "Whitespace-only version string should return empty string");
  }

  @Test
  public void testParseNodeNameFromVersionStr_MalformedVersions() {
    // Test version with wrong separator - semicolon is not a comma, so whole string becomes one
    // element
    assertEquals(
        MALFORMED_VERSION_WRONG_SEPARATOR,
        parseNodeNameFromVersionStr(MALFORMED_VERSION_WRONG_SEPARATOR),
        "Version with wrong separator should return the whole string as node name");

    // Test version that's too short but has a node name
    assertEquals(
        CRYPTAD_NODE_NAME,
        parseNodeNameFromVersionStr(MALFORMED_VERSION_SHORT),
        "Should still extract node name from short version");

    // Test version with only commas - StringTokenizer skips empty tokens, so array is empty and
    // returns null
    assertNull(parseNodeNameFromVersionStr(",,,"), "Version with only commas should return null");

    // Test version starting with comma - StringTokenizer skips empty token, so first real token is
    // "Cryptad"
    assertEquals(
        "Cryptad",
        parseNodeNameFromVersionStr(",Cryptad,1504,1.0"),
        "Version starting with comma should return first non-empty token");
  }

  @Test
  public void testParseNodeNameFromVersionStr_UnknownNodeNames() {
    // Test with unknown node names
    assertEquals(
        "UnknownNode",
        parseNodeNameFromVersionStr("UnknownNode,1504,1.0,1504"),
        "Should extract unknown node name");
    assertEquals(
        "123", parseNodeNameFromVersionStr("123,1504,1.0"), "Should extract numeric node name");

    // Test with special characters in node name
    assertEquals(
        "Node-v2.0",
        parseNodeNameFromVersionStr("Node-v2.0,1504,1.0,1504"),
        "Should extract node name with special characters");
  }

  @Test
  public void testParseNodeNameFromVersionStr_EdgeCasesAndBoundaries() {
    // Test very long node name
    String longNodeName = "VeryLongNodeNameThatExceedsNormalExpectations";
    assertEquals(
        longNodeName,
        parseNodeNameFromVersionStr(longNodeName + ",1504,1.0,1504"),
        "Should extract very long node name");

    // Test node name with spaces (even though this would be unusual)
    assertEquals(
        "Node Name",
        parseNodeNameFromVersionStr("Node Name,1504,1.0,1504"),
        "Should extract node name with spaces");

    // Test version string with trailing commas
    assertEquals(
        CRYPTAD_NODE_NAME,
        parseNodeNameFromVersionStr("Cryptad,1504,1.0,1504,"),
        "Should extract node name with trailing commas");

    // Test version string with extra components
    assertEquals(
        FRED_NODE_NAME,
        parseNodeNameFromVersionStr("Fred,0.7,1.0,1503,extra,data"),
        "Should extract node name with extra components");
  }

  @Test
  public void testParseNodeNameFromVersionStr_IntegrationWithOtherFunctions() {
    // Test that extracted node names work correctly with other functions
    String cryptadName = parseNodeNameFromVersionStr(VALID_CRYPTAD_VERSION);
    String fredName = parseNodeNameFromVersionStr(VALID_FRED_VERSION);

    assertNotNull(cryptadName, "Should extract valid Cryptad node name");
    assertNotNull(fredName, "Should extract valid Fred node name");

    // Test with compareBuildNumbers
    assertTrue(
        compareBuildNumbers(cryptadName, 1504, fredName, 1503) > 0,
        "Extracted Cryptad name should work with compareBuildNumbers");

    // Test with isBuildAtLeast
    assertTrue(
        isBuildAtLeast(cryptadName, 1, MIN_FRED_BUILD),
        "Extracted Cryptad name should work with isBuildAtLeast");
    assertTrue(
        isBuildAtLeast(fredName, MIN_FRED_BUILD, MIN_FRED_BUILD),
        "Extracted Fred name should work with isBuildAtLeast");
  }

  @Test
  public void testParseNodeNameFromVersionStr_RealWorldVersionStrings() {
    // Test with realistic version strings that might be encountered
    assertEquals(
        CRYPTAD_NODE_NAME,
        parseNodeNameFromVersionStr("Cryptad,1504,1.0,1504"),
        "Should handle typical Cryptad version");
    assertEquals(
        FRED_NODE_NAME,
        parseNodeNameFromVersionStr("Fred,0.7,1.0,1503"),
        "Should handle typical Fred version");

    // Test with version strings that might come from different Freenet forks
    assertEquals(
        "Freenet",
        parseNodeNameFromVersionStr("Freenet,0.7,1.0,1475"),
        "Should handle Freenet version");
    assertEquals(
        "Hyphanet",
        parseNodeNameFromVersionStr("Hyphanet,1.0,1.0,2000"),
        "Should handle other fork names");
  }
}

File: src/test/kotlin/network/crypta/node/VersionTest.kt
package network.crypta.node

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

/**
 * Unit tests for the Version.kt functions, focusing on the new functions:
 * - compareBuildNumbers()
 * - isBuildAtLeast()
 * - parseNodeNameFromVersionStr()
 */
class VersionTest {

  @Test
  fun compareBuildNumbers_whenComparingCryptadAndFred_expectCryptadPreferred() {
    val lowerBuild = 1
    val higherBuild = 9999
    val sameBuild = 1500

    val cryptadLowerVsFredHigher =
      compareBuildNumbers(CRYPTAD_NODE_NAME, lowerBuild, FRED_NODE_NAME, higherBuild)
    val cryptadSameVsFredSame =
      compareBuildNumbers(CRYPTAD_NODE_NAME, sameBuild, FRED_NODE_NAME, sameBuild)
    val cryptadHigherVsFredLower =
      compareBuildNumbers(CRYPTAD_NODE_NAME, higherBuild, FRED_NODE_NAME, lowerBuild)
    val fredLowerVsCryptadHigher =
      compareBuildNumbers(FRED_NODE_NAME, lowerBuild, CRYPTAD_NODE_NAME, higherBuild)
    val fredSameVsCryptadSame =
      compareBuildNumbers(FRED_NODE_NAME, sameBuild, CRYPTAD_NODE_NAME, sameBuild)
    val fredHigherVsCryptadLower =
      compareBuildNumbers(FRED_NODE_NAME, higherBuild, CRYPTAD_NODE_NAME, lowerBuild)

    assertTrue(
      cryptadLowerVsFredHigher > 0,
      "Cryptad should be newer than Fred with lower build number",
    )
    assertTrue(
      cryptadSameVsFredSame > 0,
      "Cryptad should be newer than Fred with same build number",
    )
    assertTrue(
      cryptadHigherVsFredLower > 0,
      "Cryptad should be newer than Fred with higher build number",
    )
    assertTrue(
      fredLowerVsCryptadHigher < 0,
      "Fred should be older than Cryptad with lower build number",
    )
    assertTrue(
      fredSameVsCryptadSame < 0,
      "Fred should be older than Cryptad with same build number",
    )
    assertTrue(
      fredHigherVsCryptadLower < 0,
      "Fred should be older than Cryptad with higher build number",
    )
  }

  @Test
  fun compareBuildNumbers_whenSameNodeType_expectNumericComparison() {
    val cryptadHigher = 1505
    val cryptadLower = 1504
    val fredHigher = 1503
    val fredLower = 1502

    val cryptadHigherVsLower =
      compareBuildNumbers(CRYPTAD_NODE_NAME, cryptadHigher, CRYPTAD_NODE_NAME, cryptadLower)
    val cryptadLowerVsHigher =
      compareBuildNumbers(CRYPTAD_NODE_NAME, cryptadLower, CRYPTAD_NODE_NAME, cryptadHigher)
    val cryptadSameVsSame =
      compareBuildNumbers(CRYPTAD_NODE_NAME, cryptadLower, CRYPTAD_NODE_NAME, cryptadLower)
    val fredHigherVsLower =
      compareBuildNumbers(FRED_NODE_NAME, fredHigher, FRED_NODE_NAME, fredLower)
    val fredLowerVsHigher =
      compareBuildNumbers(FRED_NODE_NAME, fredLower, FRED_NODE_NAME, fredHigher)
    val fredSameVsSame = compareBuildNumbers(FRED_NODE_NAME, fredHigher, FRED_NODE_NAME, fredHigher)

    assertTrue(cryptadHigherVsLower > 0, "Higher Cryptad build should be newer")
    assertTrue(cryptadLowerVsHigher < 0, "Lower Cryptad build should be older")
    assertEquals(0, cryptadSameVsSame, "Same Cryptad build should be equal")
    assertTrue(fredHigherVsLower > 0, "Higher Fred build should be newer")
    assertTrue(fredLowerVsHigher < 0, "Lower Fred build should be older")
    assertEquals(0, fredSameVsSame, "Same Fred build should be equal")
  }

  @Test
  fun compareBuildNumbers_whenNodeNamesNull_expectBuildNumberComparison() {
    val higherBuild = 1505
    val lowerBuild = 1504

    val higherVsLowerBothNull = compareBuildNumbers(null, higherBuild, null, lowerBuild)
    val lowerVsHigherBothNull = compareBuildNumbers(null, lowerBuild, null, higherBuild)
    val sameVsSameBothNull = compareBuildNumbers(null, lowerBuild, null, lowerBuild)
    val higherVsLowerOneNull = compareBuildNumbers(null, higherBuild, FRED_NODE_NAME, lowerBuild)
    val lowerVsHigherOneNull = compareBuildNumbers(CRYPTAD_NODE_NAME, lowerBuild, null, higherBuild)

    assertTrue(higherVsLowerBothNull > 0, "Higher build number should win with null node names")
    assertTrue(lowerVsHigherBothNull < 0, "Lower build number should lose with null node names")
    assertEquals(0, sameVsSameBothNull, "Same build number should be equal with null node names")
    assertTrue(higherVsLowerOneNull > 0, "Higher build number should win with one null node name")
    assertTrue(lowerVsHigherOneNull < 0, "Lower build number should lose with one null node name")
  }

  @Test
  fun compareBuildNumbers_whenEdgeCases_expectNumericHandling() {
    val positive = 1
    val zero = 0
    val negative = -1
    val max = Int.MAX_VALUE
    val maxMinusOne = Int.MAX_VALUE - 1

    val positiveVsZero = compareBuildNumbers(CRYPTAD_NODE_NAME, positive, CRYPTAD_NODE_NAME, zero)
    val zeroVsNegative = compareBuildNumbers(FRED_NODE_NAME, zero, FRED_NODE_NAME, negative)
    val maxVsMaxMinusOne =
      compareBuildNumbers(CRYPTAD_NODE_NAME, max, CRYPTAD_NODE_NAME, maxMinusOne)

    assertTrue(positiveVsZero > 0, "Positive should be greater than zero")
    assertTrue(zeroVsNegative > 0, "Zero should be greater than negative")
    assertTrue(maxVsMaxMinusOne > 0, "Large build number comparison should work")
  }

  @Test
  fun compareBuildNumbers_whenUnknownNodeTypes_expectBuildNumberComparison() {
    val unknown = "Unknown"
    val anotherUnknown = "AnotherUnknown"
    val someOther = "SomeOther"

    val unknownVsAnotherUnknown = compareBuildNumbers(unknown, 1505, anotherUnknown, 1504)
    val unknownVsSomeOther = compareBuildNumbers(unknown, 1505, someOther, 1504)

    assertTrue(unknownVsAnotherUnknown > 0, "Unknown node type should compare by build number")
    assertTrue(
      unknownVsSomeOther > 0,
      "Unknown vs known should compare by build number when not Cryptad/Fred",
    )
  }

  @Test
  fun isBuildAtLeast_whenCryptadNode_expectAlwaysMeetsFredMinimum() {
    val lowBuild = 1
    val zeroBuild = 0
    val negativeBuild = -1

    val lowBuildAgainstFredMin = isBuildAtLeast(CRYPTAD_NODE_NAME, lowBuild, MIN_FRED_BUILD)
    val lowBuildAgainstHighMin = isBuildAtLeast(CRYPTAD_NODE_NAME, lowBuild, 9999)
    val zeroBuildAgainstFredMin = isBuildAtLeast(CRYPTAD_NODE_NAME, zeroBuild, MIN_FRED_BUILD)
    val negativeBuildAgainstFredMin =
      isBuildAtLeast(CRYPTAD_NODE_NAME, negativeBuild, MIN_FRED_BUILD)

    assertTrue(lowBuildAgainstFredMin, "Cryptad should always meet Fred minimum build requirement")
    assertTrue(lowBuildAgainstHighMin, "Cryptad with low build should still meet Fred minimum")
    assertTrue(zeroBuildAgainstFredMin, "Cryptad with zero build should still meet Fred minimum")
    assertTrue(
      negativeBuildAgainstFredMin,
      "Cryptad with negative build should still meet Fred minimum",
    )
  }

  @Test
  fun isBuildAtLeast_whenFredNode_expectMinimumCheck() {
    @Suppress("UnnecessaryLocalVariable") val sufficientBuild = MIN_FRED_BUILD
    val higherBuild = MIN_FRED_BUILD + 1
    val insufficientBuild = MIN_FRED_BUILD - 1
    val zeroBuild = 0

    val sufficientBuildMeets = isBuildAtLeast(FRED_NODE_NAME, sufficientBuild, MIN_FRED_BUILD)
    val higherBuildMeets = isBuildAtLeast(FRED_NODE_NAME, higherBuild, MIN_FRED_BUILD)
    val insufficientBuildMeets = isBuildAtLeast(FRED_NODE_NAME, insufficientBuild, MIN_FRED_BUILD)
    val zeroBuildMeets = isBuildAtLeast(FRED_NODE_NAME, zeroBuild, MIN_FRED_BUILD)

    assertTrue(sufficientBuildMeets, "Fred node with sufficient build should meet minimum")
    assertTrue(higherBuildMeets, "Fred node with higher build should meet minimum")
    assertFalse(insufficientBuildMeets, "Fred node with insufficient build should not meet minimum")
    assertFalse(zeroBuildMeets, "Fred node with zero build should not meet minimum")
  }

  @Test
  fun isBuildAtLeast_whenNullOrUnknownNodeName_expectFredRules() {
    val unknownNode = "UnknownNode"
    val sufficientBuild = MIN_FRED_BUILD
    val insufficientBuild = MIN_FRED_BUILD - 1

    val nullSufficient = isBuildAtLeast(null, sufficientBuild, MIN_FRED_BUILD)
    val nullInsufficient = isBuildAtLeast(null, insufficientBuild, MIN_FRED_BUILD)
    val unknownSufficient = isBuildAtLeast(unknownNode, sufficientBuild, MIN_FRED_BUILD)
    val unknownInsufficient = isBuildAtLeast(unknownNode, insufficientBuild, MIN_FRED_BUILD)

    assertTrue(nullSufficient, "Null node name with sufficient build should meet minimum")
    assertFalse(nullInsufficient, "Null node name with insufficient build should not meet minimum")
    assertTrue(unknownSufficient, "Unknown node name with sufficient build should meet minimum")
    assertFalse(
      unknownInsufficient,
      "Unknown node name with insufficient build should not meet minimum",
    )
  }

  @Test
  fun isBuildAtLeast_whenBoundaryValues_expectCorrectEvaluation() {
    val minimum = 1000
    val exactMinimum = 1000
    val belowMinimum = 999
    val aboveMinimum = 1001
    val zeroMinimum = 0

    val exactMinimumMeets = isBuildAtLeast(FRED_NODE_NAME, exactMinimum, minimum)
    val belowMinimumMeets = isBuildAtLeast(FRED_NODE_NAME, belowMinimum, minimum)
    val aboveMinimumMeets = isBuildAtLeast(FRED_NODE_NAME, aboveMinimum, minimum)
    val zeroMeetsZeroMinimum = isBuildAtLeast(FRED_NODE_NAME, 0, zeroMinimum)
    val positiveMeetsZeroMinimum = isBuildAtLeast(FRED_NODE_NAME, 1, zeroMinimum)
    val negativeMeetsZeroMinimum = isBuildAtLeast(FRED_NODE_NAME, -1, zeroMinimum)

    assertTrue(exactMinimumMeets, "Exact minimum should meet requirement")
    assertFalse(belowMinimumMeets, "One below minimum should not meet requirement")
    assertTrue(aboveMinimumMeets, "One above minimum should meet requirement")
    assertTrue(zeroMeetsZeroMinimum, "Any build should meet zero minimum")
    assertTrue(positiveMeetsZeroMinimum, "Positive build should meet zero minimum")
    assertFalse(negativeMeetsZeroMinimum, "Negative build should not meet zero minimum")
  }

  @Test
  fun parseNodeNameFromVersionStr_whenValidVersions_expectNodeName() {
    val singleCryptad = "Cryptad"
    val singleFred = "Fred"

    val cryptadFromFull = parseNodeNameFromVersionStr(VALID_CRYPTAD_VERSION)
    val cryptadFromMinimal = parseNodeNameFromVersionStr(MINIMAL_CRYPTAD_VERSION)
    val fredFromFull = parseNodeNameFromVersionStr(VALID_FRED_VERSION)
    val fredFromMinimal = parseNodeNameFromVersionStr(MINIMAL_FRED_VERSION)
    val cryptadFromSingle = parseNodeNameFromVersionStr(singleCryptad)
    val fredFromSingle = parseNodeNameFromVersionStr(singleFred)

    assertEquals(CRYPTAD_NODE_NAME, cryptadFromFull, "Should extract Cryptad node name")
    assertEquals(
      CRYPTAD_NODE_NAME,
      cryptadFromMinimal,
      "Should extract Cryptad from minimal version",
    )
    assertEquals(FRED_NODE_NAME, fredFromFull, "Should extract Fred node name")
    assertEquals(FRED_NODE_NAME, fredFromMinimal, "Should extract Fred from minimal version")
    assertEquals(
      CRYPTAD_NODE_NAME,
      cryptadFromSingle,
      "Should extract node name from single component",
    )
    assertEquals(FRED_NODE_NAME, fredFromSingle, "Should extract node name from single component")
  }

  @Test
  fun parseNodeNameFromVersionStr_whenNullOrEmpty_expectNullOrEmpty() {
    val whitespaceOnly = "   "

    val nullResult = parseNodeNameFromVersionStr(null)
    val emptyResult = parseNodeNameFromVersionStr(MALFORMED_VERSION_EMPTY)
    val whitespaceResult = parseNodeNameFromVersionStr(whitespaceOnly)

    assertNull(nullResult, "Null version string should return null")
    assertNull(emptyResult, "Empty version string should return null")
    assertEquals("", whitespaceResult, "Whitespace-only version string should return empty string")
  }

  @Test
  fun parseNodeNameFromVersionStr_whenMalformedVersions_expectBestEffort() {
    val onlyCommas = ",,,"
    val leadingComma = ",Cryptad,1504,1.0"

    val wrongSeparatorResult = parseNodeNameFromVersionStr(MALFORMED_VERSION_WRONG_SEPARATOR)
    val shortVersionResult = parseNodeNameFromVersionStr(MALFORMED_VERSION_SHORT)
    val onlyCommasResult = parseNodeNameFromVersionStr(onlyCommas)
    val leadingCommaResult = parseNodeNameFromVersionStr(leadingComma)

    assertEquals(
      MALFORMED_VERSION_WRONG_SEPARATOR,
      wrongSeparatorResult,
      "Version with wrong separator should return the whole string as node name",
    )
    assertEquals(
      CRYPTAD_NODE_NAME,
      shortVersionResult,
      "Should still extract node name from short version",
    )
    assertNull(onlyCommasResult, "Version with only commas should return null")
    assertEquals(
      "Cryptad",
      leadingCommaResult,
      "Version starting with comma should return first non-empty token",
    )
  }

  @Test
  fun parseNodeNameFromVersionStr_whenUnknownNodeNames_expectFirstToken() {
    val unknownVersion = "UnknownNode,1504,1.0,1504"
    val numericVersion = "123,1504,1.0"
    val specialVersion = "Node-v2.0,1504,1.0,1504"

    val unknownResult = parseNodeNameFromVersionStr(unknownVersion)
    val numericResult = parseNodeNameFromVersionStr(numericVersion)
    val specialResult = parseNodeNameFromVersionStr(specialVersion)

    assertEquals("UnknownNode", unknownResult, "Should extract unknown node name")
    assertEquals("123", numericResult, "Should extract numeric node name")
    assertEquals("Node-v2.0", specialResult, "Should extract node name with special characters")
  }

  @Test
  fun parseNodeNameFromVersionStr_whenEdgeCases_expectNodeName() {
    val longNodeName = "VeryLongNodeNameThatExceedsNormalExpectations"
    val longNodeVersion = "$longNodeName,1504,1.0,1504"
    val spaceNodeVersion = "Node Name,1504,1.0,1504"
    val trailingCommaVersion = "Cryptad,1504,1.0,1504,"
    val extraComponentVersion = "Fred,0.7,1.0,1503,extra,data"

    val longNodeResult = parseNodeNameFromVersionStr(longNodeVersion)
    val spaceNodeResult = parseNodeNameFromVersionStr(spaceNodeVersion)
    val trailingCommaResult = parseNodeNameFromVersionStr(trailingCommaVersion)
    val extraComponentResult = parseNodeNameFromVersionStr(extraComponentVersion)

    assertEquals(longNodeName, longNodeResult, "Should extract very long node name")
    assertEquals("Node Name", spaceNodeResult, "Should extract node name with spaces")
    assertEquals(
      CRYPTAD_NODE_NAME,
      trailingCommaResult,
      "Should extract node name with trailing commas",
    )
    assertEquals(
      FRED_NODE_NAME,
      extraComponentResult,
      "Should extract node name with extra components",
    )
  }

  @Test
  @Suppress("UnnecessaryLocalVariable")
  fun parseNodeNameFromVersionStr_whenUsedWithOtherFunctions_expectCompatible() {
    val cryptadVersion = VALID_CRYPTAD_VERSION
    val fredVersion = VALID_FRED_VERSION

    val cryptadName = parseNodeNameFromVersionStr(cryptadVersion)
    val fredName = parseNodeNameFromVersionStr(fredVersion)
    val cryptadNameWorks = compareBuildNumbers(cryptadName, 1504, fredName, 1503) > 0
    val cryptadMeetsFredMinimum = isBuildAtLeast(cryptadName, 1, MIN_FRED_BUILD)
    val fredMeetsFredMinimum = isBuildAtLeast(fredName, MIN_FRED_BUILD, MIN_FRED_BUILD)

    assertNotNull(cryptadName, "Should extract valid Cryptad node name")
    assertNotNull(fredName, "Should extract valid Fred node name")
    assertTrue(cryptadNameWorks, "Extracted Cryptad name should work with compareBuildNumbers")
    assertTrue(cryptadMeetsFredMinimum, "Extracted Cryptad name should work with isBuildAtLeast")
    assertTrue(fredMeetsFredMinimum, "Extracted Fred name should work with isBuildAtLeast")
  }

  @Test
  fun parseNodeNameFromVersionStr_whenRealWorldVersions_expectNodeName() {
    val cryptadVersion = "Cryptad,1504,1.0,1504"
    val fredVersion = "Fred,0.7,1.0,1503"
    val freenetVersion = "Freenet,0.7,1.0,1475"
    val hyphanetVersion = "Hyphanet,1.0,1.0,2000"

    val cryptadResult = parseNodeNameFromVersionStr(cryptadVersion)
    val fredResult = parseNodeNameFromVersionStr(fredVersion)
    val freenetResult = parseNodeNameFromVersionStr(freenetVersion)
    val hyphanetResult = parseNodeNameFromVersionStr(hyphanetVersion)

    assertEquals(CRYPTAD_NODE_NAME, cryptadResult, "Should handle typical Cryptad version")
    assertEquals(FRED_NODE_NAME, fredResult, "Should handle typical Fred version")
    assertEquals("Freenet", freenetResult, "Should handle Freenet version")
    assertEquals("Hyphanet", hyphanetResult, "Should handle other fork names")
  }

  private companion object {
    const val CRYPTAD_NODE_NAME = "Cryptad"
    const val FRED_NODE_NAME = "Fred"
    const val MIN_FRED_BUILD = 1475

    const val VALID_CRYPTAD_VERSION = "Cryptad,1504,1.0,1504"
    const val VALID_FRED_VERSION = "Fred,0.7,1.0,1503"
    const val MINIMAL_CRYPTAD_VERSION = "Cryptad,1,1.0"
    const val MINIMAL_FRED_VERSION = "Fred,0.7,1.0,1"
    const val MALFORMED_VERSION_SHORT = "Cryptad,1504"
    const val MALFORMED_VERSION_EMPTY = ""
    const val MALFORMED_VERSION_WRONG_SEPARATOR = "Cryptad;1504;1.0"
  }
}
